src/main/scala/chisel3/util/Mux.scala

```diff
@@ -63,10 +63,22 @@ object MuxLookup {
     * @return the value found or the default if not
     */
   def apply[S <: UInt, T <: Data] (key: S, default: T, mapping: Seq[(S, T)]): T = {
-    var res = default
-    for ((k, v) <- mapping.reverse)
-      res = Mux(k === key, v, res)
-    res
+    /* If the mapping is defined for all possible values of the key, then don't use the default value */
+    val (defaultx, mappingx) = try {
+      val keySetSize = BigInt(1) << key.getWidth
+      val keyMask = keySetSize - 1
+      val distinctLitKeys = mapping.flatMap(_._1.litOption).map(_ & keyMask).distinct
+      if (distinctLitKeys.size == keySetSize) {
+        (mapping.head._2, mapping.tail)
+      } else {
+        (default, mapping)
+      }
+    } catch {
+      // This catches the case where the width of the key is unknown, i.e., key.getWidth fails
+      case _: Exception => (default, mapping)
```

Comment:
Rather than catching an Exception, should we just check if the width of the key is known?

Reply:
Otherwise, we should match on the exact Exception so that we don't mask other issues

Reply:
Unfortunately, [`getWidth` uses `throwException`](https://github.com/freechipsproject/chisel3/blob/92d88ffbe1663b4aa917c6c8b66a4fd697282410/chiselFrontend/src/main/scala/chisel3/Data.scala#L510) and that is inexact. I can just use `getOption` to avoid the problem entirely.

Reply:
Updated in the rebase.